shoebot/data/typography.py

```diff
@@ -178,7 +194,23 @@ def _pre_render(self):
         # self.layout.set_spacing(int(((self.lineheight-1)*self._fontsize)*Pango.SCALE)) #pango requires an int casting
         # we pass pango font description and the text to the pango layout
         self._pango_layout.set_font_description(self._pango_fontface)
-        self._pango_layout.set_text(self.text, -1)
+
+        if not self.markup_vars:
+            self._pango_layout.set_text(self.text, -1)
+        else:
+            # some of the specified settings require a Pango.Markup hack
+            # see https://stackoverflow.com/questions/55533312/how-to-create-a-letter-spacing-attribute-with-pycairo
+            # and https://developer.gnome.org/pango/1.46/pango-Markup.html
+
+            # we want to output something like
+            # <span letter_spacing="2048">Hello World</span>
+            markup = '<span'
+            # markup_vars was set in __init__()
+            for setting, value in self.markup_vars.items():
+                markup += f' {setting}="{self.markup_vars[setting]}"'
+            markup += f">{self.text}</span>"
+            self._pango_layout.set_markup(markup)
```

Comment:
It looks like the same kind of refactors may be possible - I'll document the process:

Ideas before starting:
- Having <span> .. </span> split at the beginning and end, while logical can be hard to follow, maybe it'll be possible to do something like f"<span {markup}>{some_data}</span>

OK, lets go...

The first thing, is this, code:
```python
for setting, value in self.markup_vars.items():
    markup += f' {setting}="{self.markup_vars[setting]}"'
```

Can be simplified by using the `value` variable it already has:
```python
for setting, value in self.markup_vars.items():
    markup += f' {setting}="{value}"'
```

Then after a little prototyping in jupyter it became obvious, this could be changed to a list comprehension, suitable for passing to `" ".join(... )` later:
```python
[f'{setting}={value}' for setting, value in self.markup_vars.items()]
['underline=1', 'underlinecolor=#eeaa00']
```

Next step is to convert it to use str.join:
```python
" ".join([f'{setting}={value}' for setting, value in self.markup_vars.items()])
'underline=1 underlinecolor=#eeaa00'
```

And then <span> can all be done in one place and hopefully the eye doesn't have to jump around as much to follow the logic:
```python
markup_styles = " ".join([f'{setting}="{value}"' for setting, value in self.markup_vars.items()])
self._pango_layout.set_markup(f"<span {markup_styles}>{self.text}</span>")
```

(not run through Black yet).

Reply:
I did more of this than I would normally to give some visibility into process, but hopefully move things to the test side sooner.
Next time I'll try and get a Mock going with Pango, then you can call text() from your test and verify it is called with the expected arguments + we should have a nice e2e test going.

Reply:
CHEERS
all done!